optuna/terminator/improvement/emmr.py

```diff
@@ -0,0 +1,390 @@
+from __future__ import annotations
+
+import math
+import sys
+from typing import cast
+from typing import TYPE_CHECKING
+import warnings
+
+import numpy as np
+
+from optuna._experimental import experimental_class
+from optuna.samplers._lazy_random_state import LazyRandomState
+from optuna.search_space import intersection_search_space
+from optuna.study import StudyDirection
+from optuna.terminator.improvement.evaluator import _compute_standardized_regret_bound
+from optuna.terminator.improvement.evaluator import BaseImprovementEvaluator
+from optuna.trial import FrozenTrial
+from optuna.trial import TrialState
+
+
+if TYPE_CHECKING:
+    import scipy.stats as scipy_stats
+    import torch
+
+    from optuna._gp import acqf
+    from optuna._gp import gp
+    from optuna._gp import prior
+    from optuna._gp import search_space as gp_search_space
+else:
+    from optuna._imports import _LazyImport
+
+    torch = _LazyImport("torch")
+    gp = _LazyImport("optuna._gp.gp")
+    acqf = _LazyImport("optuna._gp.acqf")
+    prior = _LazyImport("optuna._gp.prior")
+    gp_search_space = _LazyImport("optuna._gp.search_space")
+    scipy_stats = _LazyImport("scipy.stats")
+
+
+@experimental_class("4.0.0")
+class EMMREvaluator(BaseImprovementEvaluator):
+    """Evaluates a kind of regrets, called the Expected Minimum Model Regret(EMMR).
+
+    EMMR is an upper bound of "expected minimum simple regret" in the optimization process.
+
+    Expected minimum simple regret is a quantity that converges to zero only if the
+    optimization process has found the global optima.
+
+    For further information about expected minimum simple regret and the algorithm,
+    please refer to the following paper:
+
+    - `A stopping criterion for Bayesian optimization by the gap of expected minimum simple
+      regrets <https://proceedings.mlr.press/v206/ishibashi23a.html>`__
+
+    Args:
+        deterministic_objective:
+            A boolean value which indicates whether the objective function is deterministic.
+            Default is :obj:`False`.
+        delta:
+            A float number related to the criterion for termination. Default to 0.1.
+            For further information about this parameter, please see the aforementioned paper.
+        min_n_trials:
+            A minimum number of complete trials to compute the criterion. Default to 2.
+        seed:
+            A random seed for EMMREvaluator.
+
+    Example:
+
+        .. testcode::
+
+            import logging
```

Comment:
Not a strong opinion, but we can use `print` instead of `logging.info` to simplify the example.

Reply:
Please refer to the [code convention](https://github.com/optuna/optuna/wiki/Coding-Style-Conventions#block-and-inline-comments-are-supposed-to-be-sentences) of this repository.

Reply:
Agreed, but using `print` causes doctest failure. I have changed it to comment.